bson/src/main/org/bson/io/ByteBufferBsonInput.java

```diff
@@ -146,26 +148,70 @@ private String readString(final int size) {
         } else {
-            byte[] bytes = new byte[size - 1];
-            buffer.get(bytes);
-            byte nullByte = buffer.get();
-            if (nullByte != 0) {
-                throw new BsonSerializationException("Found a BSON string that is not null-terminated");
+            if (buffer.isBackedByArray()) {
+                int position = buffer.position();
+                int arrayOffset = buffer.arrayOffset();
+                int newPosition = position + stringSize;
+                buffer.position(newPosition);
+
+                byte[] array = buffer.array();
+                if (array[arrayOffset + newPosition - 1] != 0) {
+                    throw new BsonSerializationException("Found a BSON string that is not null-terminated");
+                }
+                return new String(array, arrayOffset + position, stringSize - 1, StandardCharsets.UTF_8);
+            } else if (stringSize > scratchBufferSize) {
+                scratchBufferSize = stringSize + (stringSize >> 1); //1.5 times the size
+                scratchBuffer = new byte[scratchBufferSize];
             }
-            return new String(bytes, StandardCharsets.UTF_8);
+
+            buffer.get(scratchBuffer, 0, stringSize);
+            if (scratchBuffer[stringSize - 1] != 0) {
+                throw new BsonSerializationException("BSON string not null-terminated");
+            }
+            return new String(scratchBuffer, 0, stringSize - 1, StandardCharsets.UTF_8);
         }
     }
 
     @Override
     public void skipCString() {
         ensureOpen();
-        boolean checkNext = true;
-        while (checkNext) {
-            if (!buffer.hasRemaining()) {
-                throw new BsonSerializationException("Found a BSON string that is not null-terminated");
+        int pos = buffer.position();
+        int length = computeCStringLength(pos);
+        buffer.position(pos + length);
+    }
+
+    public int computeCStringLength(final int prevPos) {
+        ensureOpen();
+        int pos = buffer.position();
+        int limit = buffer.limit();
+
+        if (UNALIGNED_ACCESS_SUPPORTED) {
+            int chunks = (limit - pos) >>> 3;
+            // Process 8 bytes at a time.
+            for (int i = 0; i < chunks; i++) {
+                long word = buffer.getLong(pos);
+                long mask = word - 0x0101010101010101L;
+                mask &= ~word;
+                mask &= 0x8080808080808080L;
+                if (mask != 0) {
+                    // first null terminator found in the Little Endian long
+                    int offset = Long.numberOfTrailingZeros(mask) >>> 3;
+                    // Found the null at pos + offset; reset buffer's position.
+                    return (pos - prevPos) + offset + 1;
+                }
+                pos += 8;
             }
-            checkNext = buffer.get() != 0;
         }
+
+        // Process remaining bytes one-by-one.
```

Comment:
the above comments is the common path so should accommodate the case when UNALIGNED_ACCESS_SUPPORTED is false. The above comment seems to only focus on the case when UNALIGNED_ACCESS_SUPPORTED is true and could mislead code reader.